### DISCUSSION TITLE
With nextjs, prisma and docker, how to pass DATABASE_URl after container is started

### DISCUSSION THREAD

**Discussion by ranlix** [prisma/prisma#18295] [Q&A] [answered]
Because our mysql information is managed in a centralized configuration center, I can't directly pass the DATABASE_URL as an environment variable directly into the container.
So I would like to ask if there is any solution to get the DATABASE_URL information asynchronously after the nextjs service is started and perform migration and connect to prisma？

**Comment by nurul3101** [comment, marked as answer]
Hello @ranlix 👋 

IIUC, you can't define the environment variable DATABASE_URL in the dockerfile as well, right? You want to pass DATABASE_URL without defining it anywhere? Do you get the DATABASE_URL from a different service?

**Reply by ranlix** [reply]
Hi there,

Yes, I can get DATABASE_URL from another API. But I think it's hard to do migration without both nextjs starts and no DATABASE_URL env at the same time.

**Reply by nurul3101** [reply]
@ranlix The solution I can think of is to create a script that fetches DATABASE_URL asynchronously before starting the migrations.
So, for example you can have a script that fetches DATABASE_URL and writes to the env file

Here's an example script (fetch-db-url.js) you can use:
```typescript
const fs = require('fs');
const path = require('path');
const axios = require('axios');

async function fetchDatabaseUrl() {
  try {
    const response = await axios.get('https://your-config-center-url/get-database-url');
    
    const envFile = path.join(__dirname, '.env');
    fs.writeFileSync(envFile, `DATABASE_URL=${response.data.DATABASE_URL}\n`);
    console.log('Successfully fetched and stored DATABASE_URL.');
  
  } catch (error) {
    console.error('Error fetching DATABASE_URL:', error.message);
  }
}

fetchDatabaseUrl();
```

And your package.json can invoke this script before invoking the migrations

```json
{
  "scripts": {
    "fetch-db-url": "node fetch-db-url.js",
    "migrate": "prisma migrate deploy",
    "start": "npm run fetch-db-url && npm run migrate && next start"
  }
}
```

So when the start script is invoked, first the custom script will be executed which will write to the env file and then migrate command will be invoked which can read the database url from the env file.

Would something like this work for you?

**Reply by ranlix** [reply]
> @ranlix The solution I can think of is to create a script that fetches DATABASE_URL asynchronously before starting the migrations. So, for example you can have a script that fetches DATABASE_URL and writes to the env file
> 
> Here's an example script (fetch-db-url.js) you can use:
> 
> ```ts
> const fs = require('fs');
> const path = require('path');
> const axios = require('axios');
> 
> async function fetchDatabaseUrl() {
>   try {
>     const response = await axios.get('https://your-config-center-url/get-database-url');
>     
>     const envFile = path.join(__dirname, '.env');
>     fs.writeFileSync(envFile, `DATABASE_URL=${response.data.DATABASE_URL}\n`);
>     console.log('Successfully fetched and stored DATABASE_URL.');
>   
>   } catch (error) {
>     console.error('Error fetching DATABASE_URL:', error.message);
>   }
> }
> 
> fetchDatabaseUrl();
> ```
> 
> And your package.json can invoke this script before invoking the migrations
> 
> ```json
> {
>   "scripts": {
>     "fetch-db-url": "node fetch-db-url.js",
>     "migrate": "prisma migrate deploy",
>     "start": "npm run fetch-db-url && npm run migrate && next start"
>   }
> }
> ```
> 
> So when the start script is invoked, first the custom script will be executed which will write to the env file and then migrate command will be invoked which can read the database url from the env file.
> 
> Would something like this work for you?

Thanks for the reply.
Yes, I added an `docker-entrypoint.sh` file which fetchs the database_url from api and export as a environment variable before migrate.  That's the only solution I can think of so far.

**Reply by nurul3101** [reply]
Thanks for getting back and confirming the solution that worked for you 🙌
